### PR TITLE
SIP-230 Circuit Breakers

### PR DESCRIPTION
**Pull request by dbeal-eth**
Implements [SIP-230](https://sips.synthetix.io/sips/sip-230/)

notes:
* futures contracts were updated even though there is not going to be any immediate updates to them due to immutability. This is fine because the Futures markets will use the old `ExchangeCircuitBreaker` contract, which is still technically supported
* circuit breaker will now operate separately from `SystemStatus`. This is documented in the SIP, but just wanted to make doubly sure its noted here
* Circuit breaker does not have a "backup" of checking previous oracle prices for when its in warmup period.
* some tests were removed because the functionality is no longer relevant or the cases were covered elsewhere
* `OneNetAggregatorDebtRatio` was changed beacuse the behavior of the circuit breaker has changed a little bit as per spec, and if the debt ratio is 0 it causes problems with the beraker breaking unnecessarily in tests/init cases. Furthermore, if you think about it, the `OneNetAggregatorDebtRatio` represents the number of debt shares you shuold get/have taken if you mint/burn synths. Therefore, if there are no synths in the system, we would want to initially mint *some* debt shares, making `1e27` a much more correct initial value.

### CODE REVIEW

**Review by artdgn — CHANGES_REQUESTED** [review]
Haven't had the chance for a full review but please avoid updating the futures markets contracts in any way (only the settings / manager contract can be changed - but will need to be redeployed). And yes, I saw you mentioned this would be ok in your description, I disagree.

We should not make changes to the sources to make them different from deployed contracts if they're not going to be redeployed, because this means the tests test different contracts from what's deployed.

Please also avoid updating the `Perps` contracts (because they're being refactored and it will just cause merge conflicts), assume they function as the `Futures` ones. When this is merged I will ensure Perps will use the new contract and interface before deployment.

**Comment by dbeal-eth** [comment]
> We should not make changes to the sources to make them different from deployed contracts if they're not going to be redeployed, because this means the tests test different contracts from what's deployed.

makes sense

> Please also avoid updating the Perps contracts (because they're being refactored and it will just cause merge conflicts), assume they function as the Futures ones. When this is merged I will ensure Perps will use the new contract and interface before deployment.

would you prefer if the changes be put into a separate branch for the time being? want to make sure its obvious what the intended changes would be based on the new circuit breaker

**Comment by artdgn** [comment]
> would you prefer if the changes be put into a separate branch for the time being? want to make sure its obvious what the intended changes would be based on the new circuit breaker

I think it will be simplest to leave them alone for the time being (in the context of this PR). If the old interface is supported, by the time we reach the review stage of PerpsV2 before their actual deployed, SIP-230 should be merged already and even if we forget to update the new interface, it would just be pointed out in review and fixed (and if it would be still forgotted, most of Perps V2 would have upgradable logic, so would be easy to fix even if somehow everyone forgets about it still).

And we'll need to maintain the previous interface for as long FuturesV1 lives, which will probably be for some time at least even after V2 is live.

**Comment by artdgn** [comment]
> would you prefer if the changes be put into a separate branch for the time being? want to make sure its obvious what the intended changes would be based on the new circuit breaker

I think it will be simplest to leave them alone for the time being (in the context of this PR). If the old interface is supported, by the time we reach the review stage of PerpsV2 before their actual deployed, SIP-230 should be merged already and even if we forget to update the new interface, it would just be pointed out in review and fixed (and if it would be still forgotted, most of Perps V2 would have upgradable logic, so would be easy to fix even if somehow everyone forgets about it still).

And we'll need to maintain the previous interface for as long FuturesV1 lives, which will probably be for some time at least even after V2 is live.

File: contracts/Issuer.sol [hunk 1]
```diff
@@ -109,22 +105,23 @@ contract Issuer is Owned, MixinSystemSettings, IIssuer {
...
+        newAddresses[4] = CONTRACT_SYNTHETIXDEBTSHARE;
+        newAddresses[5] = CONTRACT_FEEPOOL;
+        newAddresses[6] = CONTRACT_DELEGATEAPPROVALS;
+        newAddresses[7] = CONTRACT_REWARDESCROW_V2;
+        newAddresses[8] = CONTRACT_SYNTHETIXESCROW;
+        newAddresses[9] = CONTRACT_LIQUIDATOR;
+        newAddresses[10] = CONTRACT_LIQUIDATOR_REWARDS;
+        newAddresses[11] = CONTRACT_DEBTCACHE;
+        newAddresses[12] = CONTRACT_SYNTHREDEEMER;
+        newAddresses[13] = CONTRACT_SYSTEMSTATUS;
```
Comment (artdgn) [inline contracts/Issuer.sol:122]: nit: slither says the `systemStatus` getter (and the contract in resolver setup) can be removed

File: contracts/OneNetAggregatorDebtRatio.sol [hunk 2]
```diff
@@ -24,7 +24,9 @@ contract OneNetAggregatorDebtRatio is BaseOneNetAggregator {
             ISynthetixDebtShare(resolver.requireAndGetAddress("SynthetixDebtShare", "aggregate debt info")).totalSupply();
 
         uint result =
-            totalDebtShares == 0 ? 0 : totalIssuedSynths.decimalToPreciseDecimal().divideDecimalRound(totalDebtShares);
+            totalDebtShares == 0
+                ? 1000000000000000000000000000
```
Comment (artdgn) [inline contracts/OneNetAggregatorDebtRatio.sol:28]: why was it 0 before, and now is 1?

also please refactor as constant (and probably better to set the constant using 10**27 for clarity (and to prevent every reviewer from counting 27 zeros :)

File: contracts/CircuitBreaker.sol [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    /* ========== Mutating ========== */
+
+    /**
+     * Checks rate deviation from previous and its "invalid" oracle state (stale rate, of flagged by oracle).
+     * If its valid, set the `circuitBoken` flag and return false. Continue storing price updates as normal.
+     * Also, checks that system is not suspended currently, if it is - doesn't perform any checks, and
+     * returns last rate and the current broken state, to prevent synths suspensions during maintenance.
+     */
+    function probeCircuitBreaker(address oracleAddress, uint value) external onlyProbers returns (bool circuitBroken) {
```
Comment (artdgn) [inline contracts/CircuitBreaker.sol:98]: nit: please require oracleAddress not zero. 

Previously the caching was done for aggregators (that couldn't be zero), now this contract only operates on addresses so should do its own minimal validation in the mutative method

**Review by artdgn — CHANGES_REQUESTED** [review]
Mostly nits. Two main issues (other than the previous request for not touching futures/perps contracts for now):
- `ExchangeCircuitBreaker` wasn't touched, but I think it should be refactored to use the new `CircuitBreaker` so that there aren't two caching locations (one is used by `exchange` and the other one used by futures-v1). Having two caching location means that there are two separate breakers, and we probably should avoid that and have all flows use the same contract (for the logic itself).
- `resetLastValue` comment about not getting values from owner.

File: contracts/CircuitBreaker.sol [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    /**
+     * SIP-139
+     * resets the stored value for _lastValue for multiple currencies to the latest rate
+     * can be used to enable synths after a broken circuit happenned
+     * doesn't check deviations here, so believes that owner knows better
+     * emits LastRateOverridden
+     */
+    function resetLastValue(address[] calldata oracleAddresses, uint[] calldata values) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:124]: the idea of adding this extra parameter is so that the owner has a chance to confirm what they think the last value should be when setting the price oracle, in case when the call is executed (ex. the pdao prepares the transaction and when it executes a couple hours later the oracle is misbehaving again) the oracle price shifts out of range again the circuit remains broken. If car insurance companies are to be believed, an oracle that misbehaves once has a statistically higher chance of misbehaving again, sort of mentality.

If the value is erroneously set incorrectly by the owner, then the circuit will remain paused and the owner will have to figure out what the problem is

File: contracts/CircuitBreaker.sol [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        uint factor;
+        if (comparison > base) {
+            factor = comparison.divideDecimal(base);
+        } else {
+            factor = base.divideDecimal(comparison);
+        }
+
+        return factor >= getPriceDeviationThresholdFactor();
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:146]: good idea

**Comment by dbeal-eth** [comment]
thanks @artdgn please assume any comments I have not replied to will be addressed with a corresponding suggested change

File: contracts/OneNetAggregatorDebtRatio.sol (hunk 2 above)
Comment (dbeal-eth) [reply contracts/OneNetAggregatorDebtRatio.sol:28]: its 1 now because, after some exploration of the reasoning behind having it 0 vs 1 before and editing the tests for this PR, it makes more sense to have it as 1 because `0` will inevitably cause unnecessary divide by 0 issues when the system has no debt, and if the system goes from having no debt to having debt, the first ratio will be `1e27`, so it preserves continuity in that regard

will fix the 0s. Didnt know the `**` syntax was supported in soldiity, that is nice.

good way to see why the change is made is to just change it back and see what breaks in the tests

its an edge case that doesn't really affect the existing system

File: contracts/CircuitBreaker.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:124]: also the owner can specify `value = 0` if they want to opt out of this functionality

File: contracts/CircuitBreaker.sol [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        return false;
+    }
+
+    // ========== MODIFIERS =======
+
+    modifier onlyProbers() {
+        require(
+            msg.sender == requireAndGetAddress(CONTRACT_ISSUER) || msg.sender == requireAndGetAddress(CONTRACT_EXRATES),
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:166]: which is more important, saving gas for stakers, or saving gas for exchangers? historically, people have often complained that minting/burning costs too much gas, but afaik people don't complain about gas when exchanging (probably because most of them are dolphins or bigger and aren't bothered by more expense gas). that is the reasoning behind this ordering, though if I'm wrong about that please lmk

File: contracts/CircuitBreaker.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:146]: actually, since `priceDeviationThresholdFactor` has already been in `SystemSettings` all this time, I don't really its worth it to move the function now. The amount of gas saved might be ~10k units per probe, which comes out about less than $2 on mainnet at current ETH/gas prices, and it increases the amount of code changes and testing for this PR

File: contracts/ExchangeRates.sol [hunk 7]
```diff
@@ -68,8 +72,58 @@ contract ExchangeRates is Owned, MixinSystemSettings, IExchangeRates {
...
+        if (currencyKey == sUSD) {
+            return (rateAndTime.rate, false, false);
+        }
+
+        bool broken = circuitBreaker().probeCircuitBreaker(address(aggregators[currencyKey]), rateAndTime.rate);
+
+        return (
+            rateAndTime.rate,
+            broken,
+            _rateIsStaleWithTime(getRateStalePeriod(), rateAndTime.time) ||
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/ExchangeRates.sol]: e6f9905a

File: contracts/ExchangeRates.sol [hunk 8]
```diff
@@ -68,8 +72,58 @@ contract ExchangeRates is Owned, MixinSystemSettings, IExchangeRates {
...
+    {
+        require(currencyKey == "sUSD" || address(aggregators[currencyKey]) != address(0), "No such synth");
+
+        RateAndUpdatedTime memory rateAndTime = _getRateAndUpdatedTime(currencyKey);
+
+        if (currencyKey == sUSD) {
+            return (rateAndTime.rate, false, false);
+        }
+
+        bool broken = circuitBreaker().probeCircuitBreaker(address(aggregators[currencyKey]), rateAndTime.rate);
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/ExchangeRates.sol]: e6f9905a

File: contracts/CircuitBreaker.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply contracts/CircuitBreaker.sol:171]: actually it's not a call, so doesn't really save much either way (just an SLOAD and not a CALL)

File: contracts/CircuitBreaker.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (artdgn) [reply contracts/CircuitBreaker.sol:149]: I still think it's a simple change that will improve gas on literally every kind of transaction on L1. It doesn't looks like it requires complex testing, just a getter an setter with access control.

**Review by artdgn — CHANGES_REQUESTED** [review]
Thanks for addressing many of the things in the last review. One last important bit that I think you didn't address was the first point in https://github.com/Synthetixio/synthetix/pull/1813#pullrequestreview-1026928324 about refactoring current `ExchangeCircuitBreaker` (that will still be used by futures markets) to use the new circuit breaker, so that we don't have two separate breakers, but only one (and the old one will be proxying to the new one).

**Comment by dbeal-eth** [comment]
> Thanks for addressing many of the things in the last review. One last important bit that I think you didn't address was the first point in [#1813 (review)](https://github.com/Synthetixio/synthetix/pull/1813#pullrequestreview-1026928324) about refactoring current `ExchangeCircuitBreaker` (that will still be used by futures markets) to use the new circuit breaker, so that we don't have two separate breakers, but only one (and the old one will be proxying to the new one).

@artdgn sorry I think I missed that comment or something. That sounds like a good idea, I will see about updating `ExchangeCircuitBreaker` as you suggest

File: contracts/CircuitBreaker.sol [hunk 9]
```diff
@@ -18,6 +18,8 @@ import "./Proxyable.sol";
 // Chainlink
 import "@chainlink/contracts-0.0.10/src/v0.5/interfaces/AggregatorV2V3Interface.sol";
 
+import "hardhat/console.sol";
```
Comment (artdgn) [inline contracts/CircuitBreaker.sol:21]: whoops

**Review by artdgn — APPROVED** [review]

File: contracts/ExchangeRates.sol [hunk 10]
```diff
@@ -523,6 +575,10 @@
         return flags.getFlag(aggregator);
     }
 
+    function _rateIsCircuitBroken(bytes32 currencyKey, uint rate) internal view returns (bool) {
+        return circuitBreaker().isInvalid(address(aggregators[currencyKey]), rate);
+    }
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/ExchangeRates.sol:580]: we cant do anything about this

File: contracts/CircuitBreaker.sol [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+ *
+ * https://docs.synthetix.io/contracts/source/contracts/CircuitBreaker
+ */
+contract CircuitBreaker is Owned, MixinSystemSettings, ICircuitBreaker {
+    using SafeMath for uint;
+    using SafeDecimalMath for uint;
+
+    bytes32 public constant CONTRACT_NAME = "CircuitBreaker";
+
+    uint internal _numCircuitsBroken = 0;
```
[the remark this answers is not in the document]
Comment (dbeal-eth) [reply contracts/CircuitBreaker.sol:39]: im confused
